[clang] Add support for new loop attribute [[clang::code_align()]]

smanna12 opened the pull request:
This patch adds support for new loop attribute: [[clang::code_align(N)]].
This attribute applies to a loop and specifies the byte alignment for a loop.
The attribute accepts a positive integer constant initialization expression
indicating the number of bytes for the minimum alignment boundary.
Its value must be a power of 2, between 1 and 4096 (inclusive).

smanna12 committed:
[clang] Add support for new loop attribute [[clang::code_align()]]

AaronBallman commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,56 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+// Emit duplicate error for [[clang::code_align()]] attribute.
+template <typename LoopAttrT>
+static void
+CheckForDuplicateLoopAttribute(Sema &S,
+                               const SmallVectorImpl<const Attr *> &Attrs) {
+  const LoopAttrT *LoopAttr = nullptr;
+
+  for (const auto *I : Attrs) {
+    if (LoopAttr && isa<LoopAttrT>(I)) {
+      // Cannot specify same type of attribute twice.
+      S.Diag(I->getLocation(), diag::err_loop_attr_duplication) << LoopAttr;
+    }
+    if (isa<LoopAttrT>(I))
+      LoopAttr = cast<LoopAttrT>(I);
+  }
+}
+
```
It's probably a bit hostile for it to be an error if the values are the same, but also, I'm not certain how you'd get duplicate attributes on a statement without a copy-paste error so maybe this is fine.
```suggestion
CheckForDuplicateLoopAttribute(Sema &S,
                               const SmallVectorImpl<const Attr *> &Attrs) {
  const Attr *A = nullptr;

  for (const auto *I : Attrs) {
    if (isa<LoopAttrT>(I)) {
      if (A) {
        // Cannot specify same type of attribute twice.
        S.Diag(I->getLocation(), diag::err_loop_attr_duplication) << A;
       }
       A = I;
    }
  }
}

```

erichkeane replied:
I could definitely see macros causing duplicates, presumably as we get ones that are compound/for different functions.  I would think ones that have the same value aren't worth diagnosing on.

smanna12 replied:
I have updates codes for duplicate attribute values check https://github.com/llvm/llvm-project/pull/70762/commits/d049dc9997bdb78ff7e7cbf9b04aa42b9274cfd9

AaronBallman reviewed (commented):
Please be sure to add a reasonable description to the patch summary so that reviewers know more about what's going on, but thank you for already having written docs (that helps give context).

The changes should come with a release note as well.

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4280,3 +4280,12 @@ def PreferredType: InheritableAttr {
+                   C23<"clang", "code_align">];
+  let Subjects = SubjectList<[ForStmt, CXXForRangeStmt, WhileStmt, DoStmt],
+                              ErrorDiag, "'for', 'while', and 'do' statements">;
+  let Args = [ExprArgument<"NExpr">];
```
This needs a more descriptive name

smanna12 replied:
I have updated the Args.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7416,3 +7416,46 @@ that ``p->array`` must have at least ``p->count`` number of elements available:
+  let Category = DocCatVariable;
+  let Heading = "clang::code_align";
+  let Content = [{
+The ``clang::code_align(N)`` attribute applies to a loop and it specifies the
```
```suggestion
The ``clang::code_align(N)`` attribute applies to a loop and specifies the
```

smanna12 replied:
Done

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7416,3 +7416,46 @@ that ``p->array`` must have at least ``p->count`` number of elements available:
+byte alignment for a loop. The attribute accepts a positive integer constant
+initialization expression indicating the number of bytes for the minimum
+alignment boundary. Its value must be a power of 2, between 1 and 4096, such as
+1, 2, 4, 8, and so on. This attribute sets ``llvm.loop.align`` loop metadata
```
All of the 'such as' parts don't really add anything.  I think that is worth removing.

smanna12 replied:
Done

erichkeane commented on clang/test/Sema/code_align.c:
```diff
@@ -0,0 +1,124 @@
+}
+
+int main() {
+  code_align_dependent<8, 16, 32, -10>(); // cpp-local-note{{in instantiation of function template specialization 'code_align_dependent<8, 16, 32, -10>' requested here}}
```
Please use bookmarks to put the notes on instantiations next to the error, that way they are possible to read.

smanna12 replied:
Done

erichkeane commented on clang/test/Sema/code_align.c:
```diff
@@ -0,0 +1,124 @@
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,c-local %s
```
```suggestion
// RUN: %clang_cc1 -fsyntax-only -verify=expected,c-local %s -x c
```

smanna12 replied:
Done

erichkeane commented on clang/test/Sema/code_align.c:
```diff
@@ -0,0 +1,124 @@
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,c-local %s
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cpp-local -pedantic -x c++ -std=c++11 %s 
+
+// Add diagnostics tests for Loop attribute: [[clang::code_align()]].
```
This comment doesn't really add anything

smanna12 replied:
Done. Removed

erichkeane commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -440,6 +440,14 @@ MDNode *LoopInfo::createMetadata(
   }
 
+  // Setting clang::code_align attribute.
+  if (Attrs.CodeAlign > 0) {
```
Where is this consumed?

smanna12 replied:
Added consumption on this IR in MBBPlacement: https://github.com/llvm/llvm-project/pull/71026 by @FreddyLeaf

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7416,3 +7416,46 @@ that ``p->array`` must have at least ``p->count`` number of elements available:
+byte alignment for a loop. The attribute accepts a positive integer constant
+initialization expression indicating the number of bytes for the minimum
+alignment boundary. Its value must be a power of 2, between 1 and 4096, such as
+1, 2, 4, 8, and so on. This attribute sets ``llvm.loop.align`` loop metadata
+when it applies on a loop statement.
```
Also, what guarantees does this make?  Is this a hint, or a requirement?  Are there situations where it cannot be applied correctly?

FreddyLeaf replied:
Added consumption on this IR in MBBPlacement: https://github.com/llvm/llvm-project/pull/71026

erichkeane commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -788,6 +797,20 @@ void LoopInfoStack::push(BasicBlock *Header, clang::ASTContext &Ctx,
+  //
+  // For attribute code_align:
+  // n - 'llvm.loop.align i32 n' metadata will be emitted.
+  for (const auto *A : Attrs) {
```
Aren't you separately ensuring we don't have duplicates?  If so, you can just use `llvm::find_if`

smanna12 replied:
The codegen part has been updated using https://github.com/llvm/llvm-project/pull/70762/commits/22f66bd0199300a92ea2aefc66a2640e79b734a4

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,56 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+  const LoopAttrT *LoopAttr = nullptr;
+
+  for (const auto *I : Attrs) {
+    if (LoopAttr && isa<LoopAttrT>(I)) {
```
This is another where some sort of iteration on find_if is significantly better.

smanna12 replied:
I have updated patch using find_if https://github.com/llvm/llvm-project/pull/70762/commits/3ac064bc26a516a81b2ff36ec0dc9c55350c3e3f

smanna12 committed:
Address review comments

smanna12 commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,56 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return nullptr;
+    }
```
I have added check for 4096 upper bounds.

smanna12 commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10026,6 +10026,10 @@ def err_duplicate_case_differing_expr : Error<
+def err_loop_attr_duplication : Error<
+  "duplicate loop attribute %0">;
+def err_attribute_argument_not_power_of_two : Error<
+  "%0 attribute argument must be a constant power of two greater than zero">;
```
I have added new diagnostic to tell the user about   power of 2 with upper-bound 4096

smanna12 commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -788,6 +797,20 @@ void LoopInfoStack::push(BasicBlock *Header, clang::ASTContext &Ctx,
 
+  // Translate 'loop attributes' arguments to equivalent Attr enums.
+  // It's being handled separately from LoopHintAttrs not to support
+  // legacy GNU attributes and pragma styles.
```
I have updated the comment. Please let me know if it is still unclear.

smanna12 commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4280,3 +4280,12 @@ def PreferredType: InheritableAttr {
+
+def CodeAlign: StmtAttr {
+  let Spellings = [CXX11<"clang", "code_align">,
+                   C23<"clang", "code_align">];
```
Done

smanna12 commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7416,3 +7416,46 @@ that ``p->array`` must have at least ``p->count`` number of elements available:
+The ``clang::code_align(N)`` attribute applies to a loop and it specifies the
+byte alignment for a loop. The attribute accepts a positive integer constant
+initialization expression indicating the number of bytes for the minimum
+alignment boundary. Its value must be a power of 2, between 1 and 4096, such as
```
Done

smanna12 commented:
Thank you @erichkeane and @AaronBallman for reviews and feedbacks.

>>Please be sure to add a reasonable description to the patch summary so that reviewers know more about what's going on, but thank you for already having written docs (that helps give context).

I have added description to the patch summary.

>>The changes should come with a release note as well.

Sure, I will add a release note.

smanna12 committed:
Fix clang format errors

smanna12 committed:
Fix Lit test failure

AaronBallman commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,50 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+      return nullptr;
+    E = Res.get();
+
+    int align_value = ArgVal.getSExtValue();
+    if (align_value < CodeAlignAttr::getMinValue() ||
+        align_value > CodeAlignAttr::getMaxValue() || !ArgVal.isPowerOf2()) {
+      Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
+          << CI << CodeAlignAttr::getMinValue() << CodeAlignAttr::getMaxValue();
```
```suggestion
    int AlignValue = ArgVal.getSExtValue();
    if (AlignValue < CodeAlignAttr::getMinValue() ||
        AlignValue > CodeAlignAttr::getMaxValue() || !ArgVal.isPowerOf2()) {
      Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
          << CI << CodeAlignAttr::getMinValue() << CodeAlignAttr::getMaxValue();
```
It would be good to also pass in the alignment value we calculated as part of the diagnostic (because it can be a constant expression, the actual value may not be obvious from the code snippet).

smanna12 replied:
Thank you @AaronBallman for the suggestion. I have updated source code to pass the calculated alignment value https://github.com/llvm/llvm-project/pull/70762/commits/4c40ce27e81ae40f05aeabb513881032c04feb92

AaronBallman commented on clang/test/Sema/code_align.c:
```diff
@@ -0,0 +1,127 @@
+  [[clang::code_align(-4)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+#if __cplusplus >= 201103L
+    // cpp-local-error@+4 {{integral constant expression must have integral or unscoped enumeration type, not 'double'}}
+#else
+    // c-local-error@+2 {{integer constant expression must have integer type, not 'double'}}
+#endif
```
Rather than use `#if` to surround the comments, it's easier to understand if you use a custom `-verify=` prefix that's specific to C++11 and later.

smanna12 replied:
> Rather than use `#if` to surround the comments, it's easier to understand if you use a custom `-verify=` prefix that's specific to C++11 and later.

Thanks @AaronBallman for the suggestion about using custom 'verify'. I have updated test https://github.com/llvm/llvm-project/pull/70762/commits/4c40ce27e81ae40f05aeabb513881032c04feb92.

constexpr  and template only works with C++, so i have used '#if __cplusplus' to surround the C++ source codes otherwise it causes errors like "use of undeclared identifier 'constexpr" and "unknown type name 'template" with C flag in RUN line.

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10026,6 +10026,10 @@ def err_duplicate_case_differing_expr : Error<
+def err_loop_attr_duplication : Error<
+  "duplicate loop attribute %0">;
+def err_attribute_power_of_two_in_range : Error<
+  "%0 attribute must be a constant power of two between %1 and %2 inclusive">;
```
```suggestion
  "%0 attribute requires an integer argument which is a constant power of two between %1 and %2 inclusive%select{|; got %3}4">;
```
The idea here is that for non-integer constant expression you can say it requires the integer argument, and if you got an integer constant expression but it's out of range, then you'll add the "got <whatever>" part as well.

smanna12 replied:
I have updated diagnostic I have updated source code to pass the calculated alignment value https://github.com/llvm/llvm-project/pull/70762/commits/4c40ce27e81ae40f05aeabb513881032c04feb92

AaronBallman reviewed (commented):
Thank you for the link to the LLVM side of things, that helps!

smanna12 committed:
Address review comments

smanna12 committed:
Fix clang format errors

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10026,6 +10026,11 @@ def err_duplicate_case_differing_expr : Error<
+  "duplicate loop attribute %0">;
+def err_attribute_power_of_two_in_range : Error<
+  "%0 attribute requires an integer argument which is a constant power of two "
+  "between %1 and %2 inclusive - got %3">;
```
```suggestion
  "between %1 and %2 inclusive; got %3">;
```

smanna12 replied:
Done

AaronBallman commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -788,6 +797,17 @@ void LoopInfoStack::push(BasicBlock *Header, clang::ASTContext &Ctx,
+  // Identify loop attribute 'code_align' from Attrs.
+  // For attribute code_align:
+  // n - 'llvm.loop.align i32 n' metadata will be emitted.
+  for (const auto *A : Attrs) {
+    if (const auto *CodeAlign = dyn_cast<CodeAlignAttr>(A)) {
+      const auto *CE = cast<ConstantExpr>(CodeAlign->getAlignment());
+      llvm::APSInt ArgVal = CE->getResultAsAPSInt();
+      setCodeAlign(ArgVal.getSExtValue());
+    }
+  }
```
```suggestion
  if (const auto *CodeAlign = getSpecificAttr<CodeAlignAttr>(Attrs)) {
    const auto *CE = cast<ConstantExpr>(CodeAlign->getAlignment());
    llvm::APSInt ArgVal = CE->getResultAsAPSInt();
    setCodeAlign(ArgVal.getSExtValue());
  }
```

smanna12 replied:
I have updated codegen change to fix the build issue below:
```
if (const auto *CodeAlign = getSpecificAttr<const CodeAlignAttr>(Attrs)) {
   const auto *CE = cast<ConstantExpr>(CodeAlign->getAlignment());
   llvm::APSInt ArgVal = CE->getResultAsAPSInt();
   setCodeAlign(ArgVal.getSExtValue());
```
.../include/clang/AST/AttrIterator.h: In instantiation of ‘SpecificAttr* clang::specific_attr_iterator<SpecificAttr, Container>::operator*() const [with SpecificAttr = clang::CodeAlignAttr; Container = llvm::ArrayRef<const clang::Attr*>; clang::specific_attr_iterator<SpecificAttr, Container>::reference = clang::CodeAlignAttr*]’:
.../include/clang/AST/AttrIterator.h:120:12:   required from ‘SpecificAttr* clang::getSpecificAttr(const Container&) [with SpecificAttr = clang::CodeAlignAttr; Container = llvm::ArrayRef<const clang::Attr*>]’
.../lib/CodeGen/CGLoopInfo.cpp:803:67:   required from here
/.../include/clang/AST/AttrIterator.h:67:30: error: invalid conversion from ‘llvm::CastInfo<clang::CodeAlignAttr, const clang::Attr*, void>::CastReturnType {aka const clang::CodeAlignAttr*}’ to ‘clang::specific_attr_iterator<clang::CodeAlignAttr, llvm::ArrayRef<const clang::Attr*> >::reference {aka clang::CodeAlignAttr*}’ [-fpermissive]
     return cast<SpecificAttr>(*Current);

AaronBallman commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,57 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+      return nullptr;
+    E = Res.get();
+
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return nullptr;
+    }
```
This doesn't seem necessary as `0` is not a power of two in the correct range.
```suggestion
```

smanna12 replied:
Removed.

AaronBallman commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,57 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+    // This attribute requires an integer argument which is a constant power of
+    // two between 1 and 4096 inclusive.
+    int AlignValue = ArgVal.getSExtValue();
+    if (AlignValue > CodeAlignAttr::getMaxValue() || !ArgVal.isPowerOf2()) {
+      Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
```
```suggestion
    if (AlignValue < CodeAlignAttr::getMinValue() || AlignValue > CodeAlignAttr::getMaxValue() || !ArgVal.isPowerOf2()) {
      Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
```
You'll have to adjust for 80-col limits.

smanna12 replied:
Done

smanna12 committed:
Address review comments

smanna12 committed:
Fix lit test failure

smanna12 committed:
Fix DuplicateCodeAlignAttrs check

smanna12 committed:
Fix Clang format errors

smanna12 committed:
Remove wrong function

smanna12 committed:
Merge remote-tracking branch 'my_remote/main' into AddCodeAlignAttr

smanna12 commented:
@AaronBallman, @erichkeane I addressed comments in the PR, could you please revisit this PR? Thank you!

erichkeane commented:
> @AaronBallman, @erichkeane I have addressed comments in the PR, could you please revisit this PR? Thank you!

I'm still away at the WG21 meeting, but I've got this on my list of things to review when I return.

smanna12 commented:
> > @AaronBallman, @erichkeane I have addressed comments in the PR, could you please revisit this PR? Thank you!
> 
> I'm still away at the WG21 meeting, but I've got this on my list of things to review when I return.

Thank you @AaronBallman

smanna12 committed:
Merge remote-tracking branch 'my_remote/main' into AddCodeAlignAttr

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4297,3 +4297,19 @@ def PreferredType: InheritableAttr {
+                              ErrorDiag, "'for', 'while', and 'do' statements">;
+  let Args = [ExprArgument<"Alignment">];
+  let Documentation = [CodeAlignAttrDocs];
+  let AdditionalMembers = [{
```
These members make little sense to be here to me.  If we don't intend for them to ever be target specific, I could imagine making them static variables for the purpose of picking them up in a way that doesn't look like it uses 'magic numbers'.  If some day they might be target specific, they should likely be a part of TargetInfo or similar.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10025,6 +10025,11 @@ def err_duplicate_case_differing_expr : Error<
+  "duplicate loop attribute %0">;
+def err_attribute_power_of_two_in_range : Error<
+  "%0 attribute requires an integer argument which is a constant power of two "
+  "between %1 and %2 inclusive; got %3">;
```
we don't typically do something like `got %3` here?  I've at least not seen us do that before.  If we wish to provide it (as perhaps the person might get confused after constant evaluation?), it should probably be `provided argument was %3`

smanna12 replied:
>>I've at least not seen us do that before. If we wish to provide it (as perhaps the person might get confused after constant evaluation?), it should probably be provided argument was %3

It makes sense to me. Thanks for the suggestion.

smanna12 committed:
Update diagnostic message and test

smanna12 committed:
Merge remote-tracking branch 'my_remote/main' into AddCodeAlignAttr

smanna12 committed:
Update min and max aligment check

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -334,11 +334,12 @@ CodeAlignAttr *Sema::BuildCodeAlignAttr(const AttributeCommonInfo &CI,
     int AlignValue = ArgVal.getSExtValue();
-    if (AlignValue < CodeAlignAttr::getMinValue() ||
-        AlignValue > CodeAlignAttr::getMaxValue() || !ArgVal.isPowerOf2()) {
+    static int MaximumAlignment = 4096;
```
Ah,  I meant these still as addition members, as static-constexpr.  That way it is still `CodeAlignAttr::MinimumAlignment` and `CodeAlignAttr::MaximumAlignment`, but not as a function that implies that it is not a constant value.

smanna12 replied:
I see! Thanks @erichkeane for clarification. Fixed now with commit https://github.com/llvm/llvm-project/pull/70762/commits/aa85ed1ede892c7da3d5c2dee5886dd0e2ad0cb7

smanna12 committed:
Add additional members as static-constexpr

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,66 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+
+    // This attribute requires an integer argument which is a constant power of
+    // two between 1 and 4096 inclusive.
+    int AlignValue = ArgVal.getSExtValue();
```
this should be int64.  Also, we probably  need a test for 'larger than 64 bit' I think?  Else this is going to give weird behavior.

erichkeane replied:
To be more clear, I think SExt makes sense, but this would cause a problem for `9223372036854775808` (int64-max +1).

smanna12 replied:
with SExt, we can see this:
``
test/Sema/code_align.c:69:5: error: 'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; got -9223372036854775808
    [[clang::code_align(9223372036854775808)]]
       ^
``
With ZExt, we can see correct value:
```
uint64_t AlignValue = ArgVal.getZExtValue();
test/Sema/code_align.c:69:5: error: 'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; got 9223372036854775808
     [[clang::code_align(9223372036854775808)]]
````

smanna12 replied:
> I think SExt makes sense, but this would cause a problem for `9223372036854775808` (int64-max +1).

Yes.

@erichkeane,  I have updated patch to avoid this kind of problem with commit https://github.com/llvm/llvm-project/pull/70762/commits/088767463be990eb8af4e62a6c66cf4ba06613c3.

smanna12 committed:
Address review comments

smanna12 committed:
Fix test

smanna12 committed:
Fix lit test failures

erichkeane commented on clang/test/Sema/code_align.c:
```diff
@@ -0,0 +1,126 @@
+  [[clang::code_align(9223372036854775808)]]
+  for(int I=0; I<256; ++I) { bar(I); }
+
+  // expected-error@+1 {{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 0}}
```
This looks like a bug to me.  Why does it say 0?

smanna12 replied:
@erichkeane ,This is causing assertion instead of diagnostic
Assertion failed: getActiveBits() <= 64 && "Too many bits for uint64_t", file C:\ws\src\llvm\include\llvm/ADT/APInt.h, line 1488
https://buildkite.com/llvm-project/clang-ci/builds/6708#018bcc89-82bd-49de-8b33-1e210d1a48ad

erichkeane replied:
Right, that is more or less what I was saying before, if you're converting an `APInt` that is too large for conversion operators, you'll end up getting the wrong answer or asserting, etc.  So you have to manage how that comes out better.

smanna12 replied:
@erichkeane, i have fixed the issue with commit https://github.com/llvm/llvm-project/pull/70762/commits/8bec3c455b5f55b40a8a3d8c9180c0e3d52765fa

Now This is returning a better value:
'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 2147483647
This represents the value of the provided argument truncated to an int (instead of a 128bit float).

smanna12 committed:
Fix bug for alignment value with ((__int128_t)0x1234567890abcde0ULL << 64)

smanna12 committed:
Fix clang format errors

smanna12 committed:
update patch

smanna12 committed:
Fix return nullptr check

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,81 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+    }
+
+    if (ArgVal > CodeAlignAttr::MaximumAlignment) {
+      if (ArgVal > std::numeric_limits<int32_t>::max() &&
```
This isn't a great solution either.  This results in us lying to users.  Additionally, we have no problem with less than 64 bits.

Also note, you have the same problem with something sufficiently negative as well.  Part of the problem is we only have diagnostic overloads for `<<` up to 'long long'.  I might suggest printing the expression instead if `tryGetSExtValue` fails.

smanna12 replied:
>>This isn't a great solution either. This results in us lying to users. 

@erichkeane, do you have any suggestion about how to handle this in a bettier way?

>>Additionally, we have no problem with less than 64 bits.
Correct

>>I might suggest printing the expression instead if tryGetSExtValue fails.

What is `` tryGetSExtValue``?  Could you please elaborate? Thank you

erichkeane replied:
`https://llvm.org/doxygen/classllvm_1_1APInt.html#ae30d62aebec681aceb655de2489f12ba`

smanna12 replied:
@erichkeane , Thanks for the pointer. I have updated patch with with trySExtValue() and added test case for big negative alignment value: https://github.com/llvm/llvm-project/pull/70762/commits/5ebb42f3105186ffde314496bd9203bca5f18c45

smanna12 committed:
Update patch with trySExtValue() and add test case for big negative
alignment value

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,79 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+
+    // This attribute requires an integer argument which is a constant power of
+    // two between 1 and 4096 inclusive.
+    if (ArgVal < CodeAlignAttr::MinimumAlignment ||
```
We can end up with an expression resulting in a negative that takes more than 64 bits (say, one that evaluates to an int128), so this has the same problem as the maximum case below.

smanna12 replied:
Fixed. Thanks

smanna12 committed:
Check an expression resulting in a negative that takes more than 64 bits
for negative cases

smanna12 committed:
Fix lit test

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,90 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+    // than 64 bits otherwise print the expression.
+    if (ArgVal < CodeAlignAttr::MinimumAlignment ||
+        (ArgVal < CodeAlignAttr::MaximumAlignment && !ArgVal.isPowerOf2())) {
+      if (!ArgVal.trySExtValue()) {
```
```suggestion
      if (std::optional<int64_t> Value = ArgVal.trySExtValue()) {
```

Invert the cases here so you can use the value that you got out of the `trySExtValue`.

smanna12 replied:
@erichkeane, I tried your suggested change. But it is casing the assertion. Am i missing something? 
```
if (ArgVal < CodeAlignAttr::MinimumAlignment ||
        (ArgVal < CodeAlignAttr::MaximumAlignment && !ArgVal.isPowerOf2())) {
      if (std::optional<int64_t> Value = ArgVal.trySExtValue())
        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
            << CI << CodeAlignAttr::MinimumAlignment
            << CodeAlignAttr::MaximumAlignment << E;
      else
        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
            << CI << CodeAlignAttr::MinimumAlignment
            << CodeAlignAttr::MaximumAlignment << ArgVal.getSExtValue();
      return nullptr;
    }

    if (ArgVal > CodeAlignAttr::MaximumAlignment) {
      if (std::optional<int64_t> Value = ArgVal.trySExtValue())
        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
            << CI << CodeAlignAttr::MinimumAlignment
            << CodeAlignAttr::MaximumAlignment << E;
      else
        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
            << CI << CodeAlignAttr::MinimumAlignment
            << CodeAlignAttr::MaximumAlignment << ArgVal.getZExtValue();
      return nullptr;
    }
```

erichkeane replied:
Yes, you have to change `ArgVal.getSExtValue` in the `else` of each one back to the Expression to print instead.

smanna12 replied:
Thanks!.
Done

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,90 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+        Diag(CI.getLoc(), diag::err_attribute_power_of_two_in_range)
+            << CI << CodeAlignAttr::MinimumAlignment
+            << CodeAlignAttr::MaximumAlignment << ArgVal.getSExtValue();
+        return nullptr;
```
I'd suggest extracting the 'return nullptr' until after both conditions of the `if` above, then remove the curleys from both.

smanna12 replied:
Done

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,90 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+      }
+    }
+
+    if (ArgVal > CodeAlignAttr::MaximumAlignment) {
```
You can also make this the same as the Minimum and power of 2 above, since the contents are the same (that is, add `|| ArgVal > ...Maximum..` above.

smanna12 replied:
Done

smanna12 committed:
Fix assertion and remove duplicate codes

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,71 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+static void CheckForDuplicateCodeAlignAttrs(Sema &S,
+                                            ArrayRef<const Attr *> Attrs) {
+  // Create a list of CodeAlign attributes only.
+  SmallVector<const CodeAlignAttr *, 8> OnlyCodeAlignAttrs;
```
Sorry, just looked at this again.  What is the point of going through all of this?  I would expect a simpler algorithm to be (obviously pseudo code):


```
auto *FirstItr = std::find_if(Attrs.begin(), Attrs.end(), [](const Attr *A) { return isa<const CodeAlignAttr>(A); });
if (FirstItr == Attrs.end()) // no attributes found
  return;
auto *SecondItr = std::find_if(FirstItr + 1, Attrs.end(), [](const Attr *A) { return isa<const CodeAlignAttr>(A); });
if (SecondItr == Attrs.end()) // only 1 attribute found)
  return;

// diagnose here.

```

If you still want to diagnose EVERY duplicate, you can put the 2nd find_if in a loop there instead of as an if.

smanna12 replied:
@erichkeane, i have updated duplicate attribute values check with commit https://github.com/llvm/llvm-project/pull/70762/commits/a8b13eb4846d6f1cb41b12088aa371b3fe4edd24 in a simpler way.

>>What is the point of going through all of this?

I could not able figure out in a simple way, so ended up going through all of this while using llvm::find_if()

erichkeane replied:
Ah, so the suggestion was something like I just suggested above: Do a `find_if` on the whole collection to find the 'first' attribute.  Then do a `find_if` on the rest of the range (perhaps in a loop) to find a 2nd /further one.  See the code I put above.

smanna12 replied:
Thanks @erichkeane. This is fixed with commit https://github.com/llvm/llvm-project/pull/70762/commits/09b00c761f177c3c6ad2285fa5b7642f9c7d28cc

smanna12 replied:
Does this new commit https://github.com/llvm/llvm-project/commit/09b00c761f177c3c6ad2285fa5b7642f9c7d28cc
address the concerns ? @erichkeane , @AaronBallman

erichkeane replied:
> Because this is a statement attribute, I don't think checking for duplicates is necessary or warrants this much effort. We check for conflicting duplicates on declaration attributes because redeclarations can easily introduce additional attributes. But the same is not true for statements -- the only way to get a duplicate there is with significant effort and it seems like "first/last one wins" would be fine to document. WDYT?

Given how common it is to wrap attributes in macros, and how common it is to accidentally introduce duplicates that way, I still think there is value in diagnosing that a duplicate was ignored.

AaronBallman replied:
I don't think I've ever seen this happen outside of a test suite. Most attributes hidden behind macros are of this variety:
```
#define FOO __attribute__((foo))
```
so someone would have to do:
```
FOO FOO for(;;);
```
or similar. It's not impossible, certainly. But also not something I'm particularly worried about if it's a slog to diagnose.

erichkeane replied:
The ones I see are mostly on declarations (admittedly), but I've definitely seen something akin to:

```
#define FANCY_LOOP __attribute__((foo)) __attribute__((bar)) ANOTHER_MACRO_ATTR
#define MAKE_PARALLEL_LOOP __attribute((baz)) __attribute((bar)) ANOTHER_MACRO_ATTR2

FANCY_LOOP MAKE_PARALLEL_LOOP for(;;);

// Or even...
#define FANCY_PARALLEL_LOOP FANCY_LOOP MAKE_PARALLEL_LOOP

FANCY_PARALLLEL_LOOP
for(;;);
```

erichkeane replied:
Also note, if we do NOT diagnose duplicates, we likely have to make sure that we guarantee we pick the first (or last) lexically (likely by not adding them anyway), the `Attrs` collections have changed ordering a few times, so I would hate to try to correct something like that and cause a silent breakage.

AaronBallman replied:
Yeah, still not something I'm all that worried about, tbh. But I won't stand in the way of good diagnostic behavior!

smanna12 committed:
Fix Clang format errors

smanna12 committed:
Use Simple algorithm to handle duplicate attribute values.

smanna12 committed:
Fix clang format errors

smanna12 committed:
Fix Duplicate Attribute Values diagnostic using find_if()

mikerice1969 commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -359,16 +359,34 @@ static Attr *handleCodeAlignAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
+         return isa<const CodeAlignAttr>(A);
+      });
+
+    const Attr *FirstItrAttr =
```
Don't think we need FirstItrAttr or NextItrAttr. Couldn't you use (*FirstItr) and (*NextItr) for those?

smanna12 committed:
Apply review comments

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,61 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+  const auto *LastFoundItr = FirstItr;
+
+  while (Attrs.end() != (LastFoundItr = std::find_if(LastFoundItr + 1, Attrs.end(), FindFunc)))  {
+      S.Diag((*LastFoundItr)->getLocation(), diag::err_loop_attr_duplication) << *FirstItr;
```
Actually... this is a good point for discussion @AaronBallman : I'm now thinking perhaps we should only diagnose if we find a non-identical duplicate and diagnose it as a 'conflicting' loop attr.  WDYT?  Its a little tougher since we have to re-evaluate the expression each time, but would suppress overly noisy warnings (in cases where the two match).

AaronBallman replied:
Yeah, if the attributes resolve to the same values in each case, there's really no harm in accepting the code (we do this for declaration attributes pretty regularly). However, I'd also be fine with "last attribute wins" in the case of multiple attributes with conflicting values.

smanna12 committed:
Fix clang format errors

smanna12 commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -359,16 +359,34 @@ static Attr *handleCodeAlignAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
-        S.Diag(A->getLocation(), diag::note_previous_attribute);
-      }
-      A = I;
+    const auto *FirstItr =
```
Fixed now. Thanks

AaronBallman commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,59 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+static void
+CheckForDuplicateCodeAlignAttrs(Sema &S,
+                                const SmallVectorImpl<const Attr *> &Attrs) {
+  auto FindFunc = [](const Attr *A) { return isa<const CodeAlignAttr>(A); };
+  const auto *FirstItr = std::find_if(Attrs.begin(), Attrs.end(), FindFunc);
+
+  if (FirstItr == Attrs.end()) // no attributes found
+    return;
+
+  const auto *LastFoundItr = FirstItr;
+
+  while (Attrs.end() != (LastFoundItr = std::find_if(LastFoundItr + 1,
+                                                     Attrs.end(), FindFunc))) {
+    S.Diag((*LastFoundItr)->getLocation(), diag::err_loop_attr_duplication)
+        << *FirstItr;
+    S.Diag((*FirstItr)->getLocation(), diag::note_previous_attribute);
+  }
```
I think we could be lazier here perhaps (and tackle the suggestion from @erichkeane at the same time):
```
llvm::SmallVector<const Attr *, 2> Seen;
std::optional<llvm::APSInt> PreviouslySeen;
llvm::copy_if(Attrs, std::back_inserter(Seen), [&PreviouslySeen](const Attr *A) {
  if (const auto *CCA = dyn_cast<CodeAlignAttr>(A)) {
    const auto *CE = cast<ConstantExpr>(CCA->getAlignment());
    llvm::APSInt ArgVal = CE->getResultAsAPSInt();
    bool Ret = PreviouslySeen ? *PreviouslySeen != ArgVal : true;
    PreviouslySeen = ArgVal;
    return Ret;
  }
});
```
Then we know to diagnose if `Seen.size() > 1`.

erichkeane replied:
I mean, perhaps?  That seems worse than just doing the single iteration (`I(N + M)` vs `I(N)`).  The existing version above could just pre-save the result of the 1st one (or better, calculate it when seeing hte 1st duplicate), then compare inside the while loop.

erichkeane replied:
That is, around 368 add:
`std::optional<llvm::APSInt> FirstValue;`

Then inside the while-loop something like:

``` if (!FirstValue) FirstValue = cast<ConstantExpr>(cast<CodeAlignAttr>(*FirstItr)->getAlignment())->getResultAsAPSInt();
llvm::APSInt CurValue = cast<ConstantExpr>(cast<CodeAlignAttr>(*LastFoundItr)->getAlignment())->getResultAsAPSInt();

if (FirstValue != CurValue)
   // Diags
```

smanna12 committed:
Diagnose non-identical duplicates as a 'conflicting' loop attr and
suppress identical duplicates

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,73 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+    if (!CEA || !CEB)
+      return;
+    // Test the attribute value.
+    std::optional<llvm::APSInt> FirstValue = CEA->getResultAsAPSInt();
```
So this would only need to be an optional if it is calculated out of the loop, which, I believe is a good idea to do (and only populate it here if it hasn't been done yet).

Additionally, I think it would be a good idea to return early if we decide that the alignment expression of the first one is dependent (since we don't know what the effective size will be), and skip this loop entirely.

smanna12 replied:
Fixed now. Thanks @erichkeane

smanna12 committed:
Merge remote-tracking branch 'my_remote/main' into AddCodeAlignAttr

smanna12 committed:
Address review comments

smanna12 committed:
Fix clang format errors

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,77 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+  if (!CEA)
+    return;
+  // Test the attribute value.
+  FirstValue = CEA->getResultAsAPSInt();
```
So the idea of making this an optional is that you only do the `getResultAsAPSInt` if you NEED it below.

smanna12 replied:
Done

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,77 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+
+  while (Attrs.end() != (LastFoundItr = std::find_if(LastFoundItr + 1,
+                                                     Attrs.end(), FindFunc))) {
+    llvm::APSInt SecondValue =
```
You still have to test this for dependence, you could have 1 attribute be dependent and the other not, which would cause htis to break.  You should have something like:

```
template<int ITMPL>
[[clang::code_align(2)]][[clang::code_align(ITMPL)]]
void foo() {}

void bar() {
foo <4>();
}
```

as a test, which should show the problem (as well as reversed attributes).

smanna12 replied:
>>You still have to test this for dependence, you could have 1 attribute be dependent and the other not, which would cause htis to break.

My mistake! I have fixed the crash and added tests as well.

smanna12 committed:
Fix crash and add dependence tests

erichkeane commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -322,6 +322,81 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
+      return;
+    // Test the attribute values.
+    llvm::APSInt SecondValue = CASA->getResultAsAPSInt();
+    FirstValue = CAFA->getResultAsAPSInt();
```
```suggestion
    if (!FirstValue) FirstValue = CAFA->getResultAsAPSInt();
```

It is an optional afterall...

smanna12 committed:
Address review comments

erichkeane commented on clang/test/Sema/code_align.c:
```diff
@@ -0,0 +1,164 @@
+
+int main() {
+  code_align_dependent<8, 23, 32, -10, 64>(); // #neg-instantiation
+  bar3 <4>();
```
Why didn't this and the one below it cause a diagnostic?  This results in the two having different alignments, so shouldn't the duplicate check have caught it during instantiation?

smanna12 replied:
Yes, it is missing diagnostic!

smanna12 replied:
@erichkeane, i have fixed this diagnostic issues with commit https://github.com/llvm/llvm-project/pull/70762/commits/6778154ef0d64052f62b75955a615d542bafb799

smanna12 committed:
Fix clang format errors

smanna12 committed:
Fix diagnostic issues with conflicting attributes for template cases

smanna12 committed:
Fix clang format errors

smanna12 commented:
Thank you @erichkeane for reviews!

MaskRay commented:
@efriedma-quic
